File: note-python/shortest-path/floyd-warshall.py
import sys
import heapq


def getFile(fileName):
    path = sys.argv[0].split("/")
    path.pop()
    filePath = "/".join(path)
    file = open(filePath + "/" + fileName, "r")
    return file


file = getFile("input")

answer = 0
INF = int(1e9)

V = list(map(int, file.readline().rstrip().split()))[0]
E = list(map(int, file.readline().rstrip().split()))[0]

distance = [([INF] * V) for _ in range(V)]

for i in range(E):
    s, e, cost = list(map(int, file.readline().rstrip().split()))
    start, end = s - 1, e - 1
    distance[start][end] = cost

for i in range(V):
    distance[i][i] = 0


def floydWarshall():
    for v in range(V):
        for i in range(V):
            for j in range(V):
                distance[i][j] = min(distance[i][j], distance[i][v] + distance[v][j])


floydWarshall()

for dRow in distance:
    for dCol in dRow:
        print(dCol, end=" ")
    print()


file.close()

File: leetcode/note/floyd-warshall.js
const INF = Number(1e9);
const V = 4;
const E = [
  [1, 2, 4],
  [1, 4, 6],
  [2, 1, 3],
  [2, 3, 7],
  [3, 1, 5],
  [3, 4, 4],
  [4, 3, 2],
];
const distance = Array.from({ length: V }, () => Array.from({ length: V }, () => INF));

for (const edge of E) {
  const [s, e, cost] = edge;
  let start = s - 1;
  let end = e - 1;
  distance[start][end] = cost;
}

for (let i = 0; i < V; i++) {
  distance[i][i] = 0;
}

function floydWarshall() {
  for (let v = 0; v < V; v++) {
    for (let i = 0; i < V; i++) {
      for (let j = 0; j < V; j++) {
        distance[i][j] = Math.min(distance[i][j], distance[i][v] + distance[v][j]);
      }
    }
  }
}

floydWarshall();
console.log(distance);
